File: messenger.py
from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Dict, Iterable, List, Optional, Set


class RiskAcknowledgmentRequired(Exception):
    """Raised when a service with known risks is used without user acknowledgment."""

    def __init__(self, service: str, warning: str) -> None:
        self.service = service
        self.warning = warning
        super().__init__(f"{service!r} requires acknowledgment: {warning}")


@dataclass
class Message:
    sender: str
    recipient: str
    content: str
    protocol: str
    timestamp: datetime = field(default_factory=lambda: datetime.now(timezone.utc))

    def copy_for_protocol(self, protocol: str) -> "Message":
        return Message(
            sender=self.sender,
            recipient=self.recipient,
            content=self.content,
            protocol=protocol,
        )


class LocalStore:
    """Local-only storage to keep every message and warning on device."""

    def __init__(self) -> None:
        self.outbox: List[Dict] = []
        self.inbox: List[Dict] = []
        self.audit: List[str] = []

    @staticmethod
    def _serialize(message: Message, via: str) -> Dict:
        return {
            "sender": message.sender,
            "recipient": message.recipient,
            "content": message.content,
            "protocol": message.protocol,
            "via": via,
            "timestamp": message.timestamp.isoformat(),
        }

    def record_outgoing(self, message: Message, via: str) -> None:
        self.outbox.append(self._serialize(message, via))

    def record_incoming(self, message: Message, via: str) -> None:
        self.inbox.append(self._serialize(message, via))

    def record_audit(self, entry: str) -> None:
        self.audit.append(f"{datetime.now(timezone.utc).isoformat()} {entry}")

    def history(self) -> Dict[str, List[Dict]]:
        return {"outbox": list(self.outbox), "inbox": list(self.inbox), "audit": list(self.audit)}


class RiskRegistry:
    """Keeps a list of risks per service/protocol to warn the user before usage."""

    DEFAULT_RISKS: Dict[str, str] = {
        "sms": "SMS is not end-to-end encrypted and can be intercepted by carriers.",
        "rcs": "Carrier-managed RCS may leak metadata; E2EE depends on client and carrier support.",
        "matrix federation": "Federated homeservers can observe metadata; trust only servers you control.",
        "whatsapp": "Proprietary network; metadata and backups may be accessible to the provider.",
        "facebook messenger": "Proprietary network; metadata can be profiled.",
    }

    def __init__(self, custom_risks: Optional[Dict[str, str]] = None) -> None:
        self.risks = {**self.DEFAULT_RISKS, **(custom_risks or {})}

    @staticmethod
    def _normalize(service: str) -> str:
        return service.strip().lower()

    def get_warning(self, service: str) -> Optional[str]:
        return self.risks.get(self._normalize(service))

    def describe(self) -> Dict[str, str]:
        return dict(self.risks)


class Connector:
    """Base connector that never leaves the local process."""

    def __init__(self, name: str, protocol: str, store: LocalStore) -> None:
        self.name = name
        self.protocol = protocol
        self.store = store

    def send(self, message: Message) -> None:
        self.store.record_outgoing(message, self.name)

    def receive(self, message: Message) -> None:
        self.store.record_incoming(message, self.name)


class MatrixConnector(Connector):
    def __init__(self, store: LocalStore) -> None:
        super().__init__(name="matrix", protocol="matrix", store=store)


class RCSConnector(Connector):
    def __init__(self, store: LocalStore) -> None:
        super().__init__(name="rcs", protocol="rcs", store=store)


class ThirdPartyConnector(Connector):
    def __init__(self, name: str, store: LocalStore) -> None:
        super().__init__(name=name, protocol=name, store=store)


class MessengerService:
    """Local-only messenger that bridges Matrix/RCS and other services with warnings."""

    def __init__(self, store: Optional[LocalStore] = None, registry: Optional[RiskRegistry] = None) -> None:
        self.store = store or LocalStore()
        self.registry = registry or RiskRegistry()
        self.connectors: Dict[str, Connector] = {}
        self.bridges: Dict[str, List[str]] = {}
        self.acknowledged: Set[str] = set()

    @staticmethod
    def _normalize(service: str) -> str:
        return service.strip().lower()

    def add_connector(self, connector: Connector) -> None:
        self.connectors[self._normalize(connector.name)] = connector
        self.store.record_audit(f"connector-added {connector.name}")

    def acknowledge_service(self, service: str) -> Optional[str]:
        norm = self._normalize(service)
        warning = self.registry.get_warning(norm)
        self.acknowledged.add(norm)
        if warning:
            self.store.record_audit(f"warning-acknowledged {norm}: {warning}")
        return warning

    def _ensure_acknowledged(self, service: str) -> None:
        norm = self._normalize(service)
        warning = self.registry.get_warning(norm)
        if warning and norm not in self.acknowledged:
            raise RiskAcknowledgmentRequired(norm, warning)

    def connect_services(self, origin: str, targets: Iterable[str]) -> None:
        norm_origin = self._normalize(origin)
        self.bridges.setdefault(norm_origin, [])
        for target in targets:
            norm_target = self._normalize(target)
            if norm_target not in self.bridges[norm_origin]:
                self.bridges[norm_origin].append(norm_target)
        self.store.record_audit(f"bridge-updated {norm_origin}->{self.bridges[norm_origin]}")

    def _get_connector(self, service: str) -> Connector:
        norm = self._normalize(service)
        if norm not in self.connectors:
            raise KeyError(f"No connector registered for {service}")
        return self.connectors[norm]

    def send_message(
        self,
        service: str,
        sender: str,
        recipient: str,
        content: str,
        *,
        timestamp: Optional[datetime] = None,
    ) -> None:
        norm = self._normalize(service)
        self._ensure_acknowledged(norm)
        connector = self._get_connector(norm)
        message = Message(
            sender=sender,
            recipient=recipient,
            content=content,
            protocol=connector.protocol,
            timestamp=timestamp or datetime.now(timezone.utc),
        )
        connector.send(message)
        for target in self.bridges.get(norm, []):
            if target not in self.connectors:
                continue
            self._ensure_acknowledged(target)
            bridged = self._get_connector(target)
            bridged.send(message.copy_for_protocol(bridged.protocol))

    def receive_message(self, service: str, message: Message) -> None:
        norm = self._normalize(service)
        self._ensure_acknowledged(norm)
        connector = self._get_connector(norm)
        connector.receive(message)
        for target in self.bridges.get(norm, []):
            if target not in self.connectors:
                continue
            self._ensure_acknowledged(target)
            bridged = self._get_connector(target)
            bridged.receive(message.copy_for_protocol(bridged.protocol))

    def pending_warnings(self) -> Dict[str, str]:
        return {svc: warning for svc, warning in self.registry.describe().items() if svc not in self.acknowledged}

    def history(self) -> Dict[str, List[Dict]]:
        return self.store.history()


def demo() -> None:
    store = LocalStore()
    registry = RiskRegistry()
    messenger = MessengerService(store=store, registry=registry)
    messenger.add_connector(MatrixConnector(store))
    messenger.add_connector(RCSConnector(store))
    messenger.add_connector(ThirdPartyConnector("whatsapp", store))

    messenger.connect_services("matrix", ["rcs", "whatsapp"])

    print("Known service risks:")
    for service, warning in registry.describe().items():
        print(f"- {service}: {warning}")

    try:
        messenger.send_message("whatsapp", sender="alice", recipient="bob", content="Hi Bob from Matrix bridge")
    except RiskAcknowledgmentRequired as exc:
        print(f"⚠️  Warning for {exc.service}: {exc.warning}")
        messenger.acknowledge_service(exc.service)
        messenger.send_message("whatsapp", sender="alice", recipient="bob", content="Hi Bob from Matrix bridge (acknowledged)")

    messenger.acknowledge_service("matrix")
    messenger.acknowledge_service("rcs")
    messenger.send_message("matrix", sender="alice", recipient="carol", content="Routing via Matrix with bridge")

    print("\nLocal-only audit and history:")
    for entry in messenger.history()["audit"]:
        print(entry)
    print(messenger.history())


if __name__ == "__main__":
    demo()

File: app/src/main/java/com/securercs/messenger/core/Messenger.kt
package com.securercs.messenger.core

import java.time.Instant
import java.time.format.DateTimeFormatter
import java.util.Locale

class RiskAcknowledgmentRequired(val service: String, val warning: String) :
    IllegalStateException("$service requires acknowledgment: $warning")

data class Message(
    val sender: String,
    val recipient: String,
    val content: String,
    val protocol: String,
    val timestamp: Instant = Instant.now(),
) {
    fun copyForProtocol(protocol: String): Message = copy(protocol = protocol)
}

data class StoredMessage(
    val sender: String,
    val recipient: String,
    val content: String,
    val protocol: String,
    val via: String,
    val timestamp: String,
)

data class HistorySnapshot(
    val outbox: List<StoredMessage>,
    val inbox: List<StoredMessage>,
    val audit: List<String>,
)

data class SendRequest(
    val service: String,
    val sender: String,
    val recipient: String,
    val content: String,
    val targets: List<String> = emptyList(),
)

class LocalStore {
    private val formatter = DateTimeFormatter.ISO_INSTANT
    private val outbox = mutableListOf<StoredMessage>()
    private val inbox = mutableListOf<StoredMessage>()
    private val audit = mutableListOf<String>()

    private fun serialize(message: Message, via: String): StoredMessage =
        StoredMessage(
            sender = message.sender,
            recipient = message.recipient,
            content = message.content,
            protocol = message.protocol,
            via = via,
            timestamp = formatter.format(message.timestamp),
        )

    fun recordOutgoing(message: Message, via: String) {
        outbox.add(serialize(message, via))
    }

    fun recordIncoming(message: Message, via: String) {
        inbox.add(serialize(message, via))
    }

    fun recordAudit(entry: String) {
        audit.add("${formatter.format(Instant.now())} $entry")
    }

    fun history(): HistorySnapshot = HistorySnapshot(
        outbox = outbox.toList(),
        inbox = inbox.toList(),
        audit = audit.toList(),
    )
}

class RiskRegistry(customRisks: Map<String, String>? = null) {
    private val risks: Map<String, String> = DEFAULT_RISKS + (customRisks ?: emptyMap())

    fun getWarning(service: String): String? = risks[normalize(service)]

    fun describe(): Map<String, String> = risks.toMap()

    private fun normalize(service: String): String = service.trim().lowercase(Locale.getDefault())

    companion object {
        val DEFAULT_RISKS: Map<String, String> = mapOf(
            "sms" to "SMS is not end-to-end encrypted and can be intercepted by carriers.",
            "rcs" to "Carrier-managed RCS may leak metadata; E2EE depends on client and carrier support.",
            "matrix" to "Federated Matrix homeservers can observe metadata; trust only servers you control.",
            "matrix federation" to "Federated homeservers can observe metadata; trust only servers you control.",
            "whatsapp" to "Proprietary network; metadata and backups may be accessible to the provider.",
            "facebook messenger" to "Proprietary network; metadata can be profiled.",
        )
    }
}

open class Connector(
    val name: String,
    val protocol: String,
    private val store: LocalStore,
) {
    open fun send(message: Message) {
        store.recordOutgoing(message, name)
    }

    open fun receive(message: Message) {
        store.recordIncoming(message, name)
    }
}

class MatrixConnector(store: LocalStore) : Connector(name = "matrix", protocol = "matrix", store = store)
class RCSConnector(store: LocalStore) : Connector(name = "rcs", protocol = "rcs", store = store)
class ThirdPartyConnector(name: String, store: LocalStore) : Connector(name = name, protocol = name, store = store)

class MessengerService(
    private val store: LocalStore,
    private val registry: RiskRegistry,
) {
    private val connectors: MutableMap<String, Connector> = mutableMapOf()
    private val bridges: MutableMap<String, MutableList<String>> = mutableMapOf()
    private val acknowledged: MutableSet<String> = mutableSetOf()

    private fun normalize(service: String): String = service.trim().lowercase(Locale.getDefault())

    fun addConnector(connector: Connector) {
        connectors[normalize(connector.name)] = connector
        store.recordAudit("connector-added ${connector.name}")
    }

    fun services(): List<String> = connectors.keys.sorted()

    fun acknowledgeService(service: String): String? {
        val norm = normalize(service)
        val warning = registry.getWarning(norm)
        acknowledged.add(norm)
        warning?.let {
            store.recordAudit("warning-acknowledged $norm: $it")
        }
        return warning
    }

    private fun ensureAcknowledged(service: String) {
        val norm = normalize(service)
        val warning = registry.getWarning(norm)
        if (warning != null && norm !in acknowledged) {
            throw RiskAcknowledgmentRequired(norm, warning)
        }
    }

    private fun connectorFor(service: String): Connector {
        val norm = normalize(service)
        return connectors[norm] ?: error("No connector registered for $service")
    }

    fun connectServices(origin: String, targets: List<String>) {
        val normOrigin = normalize(origin)
        val stored = bridges.getOrPut(normOrigin) { mutableListOf() }
        targets.map(::normalize).forEach { target ->
            if (target !in stored) stored.add(target)
        }
        store.recordAudit("bridge-updated $normOrigin->$stored")
    }

    fun sendMessage(
        service: String,
        sender: String,
        recipient: String,
        content: String,
        targets: List<String> = emptyList(),
    ) {
        val norm = normalize(service)
        ensureAcknowledged(norm)
        val connector = connectorFor(norm)
        val message = Message(sender = sender, recipient = recipient, content = content, protocol = connector.protocol)
        connector.send(message)
        bridges[norm].orEmpty().plus(targets.map(::normalize)).distinct().forEach { target ->
            if (!connectors.containsKey(target)) return@forEach
            ensureAcknowledged(target)
            connectorFor(target).send(message.copyForProtocol(connectorFor(target).protocol))
        }
    }

    fun receiveMessage(service: String, message: Message) {
        val norm = normalize(service)
        ensureAcknowledged(norm)
        val connector = connectorFor(norm)
        connector.receive(message)
        bridges[norm].orEmpty().forEach { target ->
            if (!connectors.containsKey(target)) return@forEach
            ensureAcknowledged(target)
            connectorFor(target).receive(message.copyForProtocol(connectorFor(target).protocol))
        }
    }

    fun pendingWarnings(): Map<String, String> =
        registry.describe().filter { (service, _) ->
            val norm = normalize(service)
            norm !in acknowledged && (connectors.keys.contains(norm) || bridges.values.flatten().contains(norm))
        }

    fun history(): HistorySnapshot = store.history()
}
